crates/wasm_interp/src/tests/test_mem.rs

```diff
@@ -49,6 +49,72 @@ fn test_growmemory() {
     assert_eq!(state.memory.len(), 5 * MemorySection::PAGE_SIZE as usize);
 }
 
+#[test]
+fn test_memory_fill() {
+    let arena = Bump::new();
+    let mut module = WasmModule::new(&arena);
+
+    let pages = 3;
+    let pc = 0;
+    module.memory = MemorySection::new(&arena, pages * MemorySection::PAGE_SIZE);
+
+    const SIZE: i32 = 16;
+    let byte_value = 0xAA;
+    let destination = 0x4;
+
+    let bytes = [OpCode::MEMORY as u8, 11, 0x0];
+    module.code.bytes.extend(bytes);
+
+    let mut state = Instance::new(&arena, pages, pc, [], DefaultImportDispatcher::default());
+
+    state.value_store.push(Value::I32(destination));
+    state.value_store.push(Value::I32(byte_value));
+    state.value_store.push(Value::I32(SIZE));
+
+    // before the instruction, the memory is all zeros
+    let slice = &state.memory[destination as usize..][..SIZE as usize];
+    assert_eq!(slice, &[0; SIZE as usize]);
+
+    state.execute_next_instruction(&module).unwrap();
+
+    let slice = &state.memory[destination as usize..][..SIZE as usize];
+    assert_eq!(slice, &[byte_value as u8; SIZE as usize])
+}
+
+#[test]
+fn test_memory_copy() {
+    let arena = Bump::new();
+    let mut module = WasmModule::new(&arena);
+
+    let pages = 3;
+    let pc = 0;
+    module.memory = MemorySection::new(&arena, pages * MemorySection::PAGE_SIZE);
+
+    const SIZE: i32 = 4;
+    let source = 0x4;
+    let destination = 0x8;
+
+    let bytes = [OpCode::MEMORY as u8, 10, 0x0, 0x0];
+    module.code.bytes.extend(bytes);
+
+    let mut state = Instance::new(&arena, pages, pc, [], DefaultImportDispatcher::default());
+
+    state.value_store.push(Value::I32(destination));
+    state.value_store.push(Value::I32(source));
+    state.value_store.push(Value::I32(SIZE));
+
+    // before the instruction, the memory is all zeros
+    let slice = &mut state.memory[source as usize..][..SIZE as usize];
+    assert_eq!(slice, &[0; SIZE as usize]);
+
+    slice.fill(0xAA);
```

Comment:
~~I don't quite follow what's happening here.
To me it looks like this line is forcing the test to pass even if `execute_next_instruction` does nothing?~~

Reply:
Oh, I finally noticed that it is only filling half the memory!!!!
Wow. OK so this code is correct but it took me a couple of re-readings to figure out that it is correct!
I think some simple name changes would make it clear even when skim-reading.
It would be really helpful to have separate names like `source_slice` and `dest_slice` for the two slices. I'm not a fan of reusing the same name for two things that are already easily confused! That's the exact case where separate names are helpful.